### ISSUE TITLE
Lanczos Restarting question

### ISSUE THREAD

**Issue by peekxc** [yixuan/spectra#146]
Hi @yixuan, 

I'm trying to understand the Lanczos method, mostly via Golub & Van Loan's [Matrix Computations](https://books.google.com/books?id=5U-l8U3P-VUC&lpg=PP1&ots=7-HCRjYMan&dq=matrix%20computations&lr&pg=PP1#v=onepage&q=matrix%20computations&f=false) and various other articles. 

Following [this breakdown](https://slepc.upv.es/documentation/reports/str5.pdf
), my understanding is that the basic Lanczos (Algorithm 1) can compute the full spectrum of eigenvalues of a (n x n) sparse symmetric matrix using just O(n) workspace memory via the three-term recurrence, but only in exact arithmetic. That reference calls the 3-term-recurrence approach _local orthogonalization_. 

Rounding errors that occur in the eigenvalue estimation procedure are due to loss of orthogonality in the Lanczos vectors, so we must either use more Lanczos vectors or some other means of retaining orthogonality using finite-precision arithmetic. Reorthogonalization via every Lanczos vector is possible, but this implies an O(n^2) memory overhead per-iteration. Other approaches seem to include, but are not limited to, block-Lanczos methods which effectively use O(nb) memory, polynomial filtering schemes, and the so-called implicit restart scheme used here. 

I'm wondering whether it's possible to use Spectra (or ARPACK) to do either an explicit restart-like scheme, similar to Algorithm 4 from [here](https://slepc.upv.es/documentation/reports/str5.pdf), or possibly the O(nb) block Lanczos approach. Essentially, I want have the ability to fix the number of Lanczos vectors used to acquire the Ritz values. Both Spectra and e.g. the [SciPy port of ARPACK](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html#scipy.sparse.linalg.eigsh) give this as a parameter option `ncv`, but it must be larger than the number of requested eigenvalues! I'm would like to be able to fix it at some value `k` at the sacrifice of needing more iterations to obtain the spectral information up to the requisite precision. 

Is this possible to do in Spectra, possibly via some loop that uses a shift-and-invert solver?

**Comment by yixuan** [comment]
Hi @peekxc, sorry for the late reply. I was not able to get the time looking at the issues here.

Let me try to understand your question. Basically you want to compute **all** eigenvalues of a sparse matrix with only $O(nk)$ memory use. Is that correct?

**Comment by peekxc** [comment]
Yes. I was confused at the time as the material I was reading suggested this was possible, but none of the Lanczos software I could find seemed to support this, which I thought seemed like a huge discrepancy.

I ended up finding [PRIMME](https://www.cs.wm.edu/~andreas/software/doc/pyeigsh.html) actually achieves this with variations of the Jacobi-Davidson method; you can set parameters like [maxBasisSize](https://www.cs.wm.edu/~andreas/software/doc/appendix.html#c.primme_params.maxBasisSize)

**Comment by yixuan** [comment]
Right. The reason is that implicitly restarted Arnoldi/Lanczos algorithm is a Krylov space method, and the restart needs to maintain the upper Hessenberg (or tridiagonal in the Lanczos case) structure of the $H$ matrix such that $AV\approx VH$. As a result, extra space is needed to compute the "unwanted eigenvalues", i.e., the shifts for the QR iteration.

Jacobi-Davidson no longer maintains the upper Hessenberg structure of the $H$ matrix, so its restart is easy. The cost is that $H$ is now a general matrix, thus it is more time-consuming to compute the Schur decomposition.

In my experience, the `ncv` parameter is typically at the same order of `k`, so it does not result in much larger memory footprint. The real problem is to compute all eigenvalues using iterative methods like Lanczos and Jacobi-Davidson. In my opinion, they are designed to compute just a few eigenvalues.

**Comment by peekxc** [comment]
> The real problem is to compute all eigenvalues using iterative methods like Lanczos and Jacobi-Davidson. In my opinion, they are designed to compute just a few eigenvalueMs.

Right I realize most people just require a few eigenvalues but I encountered a situation where I needed to estimate *all* of the non-zero eigenvalues (but not the eigenvectors) of a nearly-full-rank `LinearOperator` which admitted a highly structured `matvec` operation (e.g. its a Laplacian type operation, $O(n)$ to evaluate). 

>  In my opinion, they are designed to compute just a few eigenvalues.

Yes I've had quite a difficult time figuring out how to adjust the hyper-parameters to keep them competitive time-wise with direct methods or with the ARPACK implementation (e.g. havent had any luck with preconditioning, really). Nonetheless, it seems if the `matvec` is cheap enough and the restart parameters are configured well, I can get all of the eigenvalues in a time thats within a small constant factor of `scipy.sparse.linalg.eigh`
